### PR TITLE
Implements plugin: channel-safely v1.2.3

### LINKED ISSUE #2818: Channel-safely cannot set marker mode
A player reported bug on discord regarding channel-safely not working correctly and crashing. Details on the crash are unclear and may very well be unrelated to channel-safely. The main trouble identified and described the player was marker mode not being set quickly enough for designations. Of course upon load or enabling the plugin the marker mode should be set for designations that require it.

I was able to reproduce the issue using the save game provided by the player. Since marker mode is part of the manager class I enabled the trace logs for the manager via `debugfilter set Trace channel manager`. Additionally also adding the debug logs for the groups processing in order to have access to some of the coordinates being processed.

The manager logs are as follows:
```
managing group #3
cave-in evaluated false
cave-in evaluated false
cave-in evaluated false
cave-in evaluated false
manage_group() is done
```

Group 3:
```
  group 3 (size: 4)
   (123,29,124)
   (123,30,124)
   (123,31,124)
   (124,31,124)
```

The manager logs are as deep as the log line seen here:
https://github.com/DFHack/dfhack/blob/a7b5ce418e931fdf19dd6662385828173d186cbf/plugins/channel-safely/channel-manager.cpp#L110-L116

This code of course calls `manage_one` after printing the log line to indicate where the algorithm is. So taking a look at that function we can see it should print another log line so long as the position passed is a valid one.
https://github.com/DFHack/dfhack/blob/a7b5ce418e931fdf19dd6662385828173d186cbf/plugins/channel-safely/channel-manager.cpp#L153-L155

So then, looking at the function used to check whether that is the case or not. We can see it checks if `IsValid()` then simply checks if the coordinates are within the bounds of the map.

https://github.com/DFHack/dfhack/blob/a7b5ce418e931fdf19dd6662385828173d186cbf/library/modules/Maps.cpp#L193-L202

https://github.com/DFHack/dfhack/blob/a7b5ce418e931fdf19dd6662385828173d186cbf/library/modules/Maps.cpp#L104-L107

Looking at both these functions, the data being compared against (other than what the plugin contains) appears to all be within `world->map`. So logging the data inside we see:
```
df.global.world.map: <world.T_map: 00007FF7646B7EA8>
| map_blocks: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ <vector<map_block*>[11664]: 00007FF7646B7EA8>
| block_index: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  <map_block***: 000001DC678B5810>
| map_block_columns: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  <vector<map_block_column*>[81]: 00007FF7646B7EC8>
| column_index: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ <map_block_column**: 000001DC678B2110>
| x_count_block: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  9
| y_count_block: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  9
| z_count_block: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  144
| x_count: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  144
| y_count: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  144
| z_count: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  144
| region_x: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ 147
| region_y: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ 208
| region_z: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ 1
| distance_lookup: ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~  <int16_t[53][]: 00007FF7646B7F0C>
```

We can see that `block_index` appears to be not `NULL` nor do the coordinates of group 3 fall outside the ranges here. Which of course means that it does not make a lot of sense for this check to fail. As the coordinate data is valid and the map appears valid.

- Nothing in the algorithm modifies the coordinates.
- The coordinates are valid
- The coordinates are output to console showing that they can be programmatically accessed and iterated over

### PR DESCRIPTION
**Pull request by cppcooper**
- Revises a few log lines
- Adds d_assert macro to replace assert usage
  - provides a log line printed to the console to indicate d_assert failed in Release/Debug builds
  
This resolves #2818 (presumably)

### CODE REVIEW

File: plugins/channel-safely/channel-manager.cpp [hunk 1]
```diff
@@ -5,7 +5,17 @@
 #include <modules/EventManager.h> //hash function for df::coord
 #include <df/block_square_event_designation_priorityst.h>
 
+#define NUMARGS(...) std::tuple_size<decltype(std::make_tuple(__VA_ARGS__))>::value
+#define d_assert(condition, ...) \
+            static_assert(NUMARGS(__VA_ARGS__) >= 1, "d_assert(condition, format, ...) requires at least up to format as arguments"); \
+            if (!condition) {                                                   \
+                DFHack::Core::getInstance().getConsole().print(__VA_ARGS__);    \
```
Comment (myk002) [inline plugins/channel-safely/channel-manager.cpp:12]: this should probably be `ERROR(manager).print(__VA_ARGS__)'`

**Review by myk002 — COMMENTED** [review]
How does this PR resolve the linked issue, though? All I see is instrumentation.

**Comment by cppcooper** [comment]
> How does this PR resolve the linked issue, though? All I see is instrumentation.

the only logical conclusion I was able to make was that the function call was never made as all the evidence pointed to the valid tile pos function being able to return true yet inexplicably returning false instead. So by revising the assertions in the manager with that d_assert the function should now always be invoked.

I'm going to test out the github windows artifact tomorrow, hopefully. I'm pretty sure this will fix the problem though.

File: plugins/channel-safely/channel-manager.cpp (hunk 1 above)
Comment (cppcooper) [reply plugins/channel-safely/channel-manager.cpp:12]: I was going to go with printerr like lethosor suggested on discord. Should I prefer this?

Comment (myk002) [reply plugins/channel-safely/channel-manager.cpp:12]: looking at the code, no, not yet. use printerr so it goes to stderr. we need to modify the debug logging framework to also print to stderr for WARN and above

Comment (cppcooper) [reply plugins/channel-safely/channel-manager.cpp:12]: Earlier today I had a brief thought about a plugin that would share stdout and stderr with a file log. I vaguely recall having written some code to do that, pretty sure it was some C code

**Comment by cppcooper** [comment]
> I'm pretty sure this will fix the problem though.

verified

File: plugins/channel-safely/channel-manager.cpp (hunk 1 above)
Comment (myk002) [reply plugins/channel-safely/channel-manager.cpp:12]: will you change this to printerr for this PR? it seems like output splitting would be a larger project

Comment (cppcooper) [reply plugins/channel-safely/channel-manager.cpp:12]: Yea, my local has the commit. I was just waiting to boot back to linux to push

**Pull request merged**